File: bindings/swift/Sources/IDAX/Microcode.swift
internal import CIDAX

/// Post-hoc microcode snapshot API.
///
/// `Microcode` is the value-typed counterpart of the live `MicrocodeContext`
/// filter API exposed in ``Decompiler``: rather than intercepting
/// instructions while Hex-Rays lifts them, it produces a deep-copied view of
/// the entire `mba_t` once lifting finishes. Snapshots are self-contained
/// and outlive any SDK state.
///
/// Typical use is driving an external IR consumer (e.g. an MIR-style SSA
/// importer for the Swift decompiler project) that wants a stable, off-line
/// view of the decompiler's microcode.
public enum Microcode {

    /// Maturity at which to snapshot the microcode.
    ///
    /// Mirrors `ida::microcode::Maturity` / SDK `mba_maturity_t`.
    /// ``lvars`` is closest to ctree input and the recommended choice for
    /// MIR-style consumers.
    public enum Maturity: Int32, Sendable, CaseIterable {
        case generated       = 1
        case preoptimized    = 2
        case locopt          = 3
        case calledArguments = 4
        case glbopt1         = 5
        case glbopt2         = 6
        case glbopt3         = 7
        case lvars           = 8
    }

    /// Basic-block kind, mirroring SDK `mblock_type_t`.
    public enum BlockKind: Int32, Sendable {
        case none     = 0
        case stop     = 1
        case noWay    = 2
        case oneWay   = 3
        case twoWay   = 4
        case nWay     = 5
        case external = 6
    }

    /// Microcode operand kind, mirroring SDK `mopt_t`.
    public enum OperandKind: Int32, Sendable {
        case none              = 0
        case register          = 1
        case numericConstant   = 2
        case stringLiteral     = 3
        case nestedInstruction = 4
        case stackVariable     = 5
        case globalAddress     = 6
        case blockReference    = 7
        case callInfo          = 8
        case localVariable     = 9
        case addressOf         = 10
        case helper            = 11
        case cases             = 12
        case floatConstant     = 13
        case registerPair      = 14
        case scattered         = 15
    }

    /// Deep-copied microcode operand snapshot.
    public struct Operand: Sendable {
        public let kind: OperandKind
        public let byteWidth: Int
        public let registerID: Int
        public let secondRegisterID: Int
        public let numericValue: Int64
        public let floatValue: Double
        public let stackOffset: Int64
        public let globalAddress: Address
        public let localVariableIndex: Int
        public let localVariableOffset: Int64
        public let helperName: String
        public let stringLiteral: String
        public let blockIndex: Int
        public let nestedInstructionID: Int
        public let ssaVersion: Int?
        public let operandProperties: UInt8
    }

    /// Deep-copied microcode instruction snapshot.
    public struct Instruction: Sendable {
        public let id: Int
        public let sourceAddress: Address
        public let opcode: Int
        public let opcodeName: String
        public let left: Operand
        public let right: Operand
        public let destination: Operand
        public let flags: UInt32
    }

    /// Deep-copied microcode basic block snapshot.
    public struct Block: Sendable {
        public let index: Int
        public let startAddress: Address
        public let endAddress: Address
        public let kind: BlockKind
        public let flags: UInt32
        public let predecessorIndices: [Int]
        public let successorIndices: [Int]
        public let instructions: [Instruction]
    }

    /// Move-only handle to a microcode snapshot.
    ///
    /// `deinit` releases the underlying C++ snapshot. Once decoded into the
    /// value types above, the result is fully `Sendable` and can be passed
    /// across actors freely; the handle itself is not.
    public struct FunctionSnapshot: ~Copyable, @unchecked Sendable {
        let handle: IdaxMicrocodeSnapshotHandle

        init(_ handle: IdaxMicrocodeSnapshotHandle) {
            self.handle = handle
        }

        deinit {
            idax_microcode_snapshot_free(handle)
        }

        public var functionAddress: Address {
            get throws(IDAError) {
                try withOutput("microcode.functionAddress", UInt64(0)) {
                    idax_microcode_snapshot_function_address(handle, $0)
                }
            }
        }

        public var maturity: Maturity {
            get throws(IDAError) {
                var raw: Int32 = 0
                try checkStatus(
                    idax_microcode_snapshot_maturity(handle, &raw),
                    "microcode.maturity"
                )
                return Maturity(rawValue: raw) ?? .lvars
            }
        }

        public var localVariablesSize: Int64 {
            get throws(IDAError) {
                try withOutput("microcode.localVariablesSize", Int64(0)) {
                    idax_microcode_snapshot_local_variables_size(handle, $0)
                }
            }
        }

        public var savedRegistersSize: Int64 {
            get throws(IDAError) {
                try withOutput("microcode.savedRegistersSize", Int64(0)) {
                    idax_microcode_snapshot_saved_registers_size(handle, $0)
                }
            }
        }

        public var stackSize: Int64 {
            get throws(IDAError) {
                try withOutput("microcode.stackSize", Int64(0)) {
                    idax_microcode_snapshot_stack_size(handle, $0)
                }
            }
        }

        public var blockCount: Int {
            get throws(IDAError) {
                var out: Int = 0
                try checkStatus(
                    idax_microcode_snapshot_block_count(handle, &out),
                    "microcode.blockCount"
                )
                return out
            }
        }

        /// Materialise all basic blocks in `mba->natural` order.
        ///
        /// Each block is decoded into Swift value types; the underlying
        /// snapshot continues to be owned by this handle.
        public var blocks: [Block] {
            get throws(IDAError) {
                let count = try blockCount
                var result: [Block] = []
                result.reserveCapacity(count)
                for index in 0..<count {
                    result.append(try block(at: index))
                }
                return result
            }
        }

        /// Materialise the block at the given zero-based index.
        public func block(at index: Int) throws(IDAError) -> Block {
            var raw = IdaxMicrocodeSnapshotBlock()
            try checkStatus(
                idax_microcode_snapshot_block(handle, index, &raw),
                "microcode.block"
            )
            defer { idax_microcode_snapshot_block_free(&raw) }
            return Block(raw: raw)
        }

        /// Resolve a nested microinstruction referenced from an operand.
        ///
        /// `id` is the value of ``Microcode/Operand/nestedInstructionID``
        /// (which is `-1` when the operand does not refer to a nested
        /// instruction). Negative ids return a `notFound` error.
        public func nestedInstruction(id: Int) throws(IDAError) -> Instruction {
            var raw = IdaxMicrocodeSnapshotInstruction()
            try checkStatus(
                idax_microcode_snapshot_nested_instruction(handle, Int32(id), &raw),
                "microcode.nestedInstruction"
            )
            defer { idax_microcode_snapshot_instruction_free(&raw) }
            return Instruction(raw: raw)
        }

        /// Materialise the local-variable table visible at this maturity.
        ///
        /// Fully populated from `.lvars`; lower maturities expose whatever
        /// the SDK has so far.
        public var localVariables: [LocalVariable] {
            get throws(IDAError) {
                var pointer: UnsafeMutablePointer<IdaxLocalVariable>? = nil
                var count: Int = 0
                try checkStatus(
                    idax_microcode_snapshot_local_variables(handle, &pointer, &count),
                    "microcode.localVariables"
                )
                defer { idax_decompiled_variables_free(pointer, count) }
                guard let pointer, count > 0 else { return [] }
                let buffer = UnsafeBufferPointer(start: pointer, count: count)
                return buffer.map { v in
                    LocalVariable(
                        name: borrowCString(v.name),
                        typeName: borrowCString(v.type_name),
                        isArgument: v.is_argument != 0,
                        width: Int(v.width),
                        hasUserName: v.has_user_name != 0,
                        storage: VariableStorage(rawValue: Int(v.storage)) ?? .unknown,
                        comment: borrowCString(v.comment)
                    )
                }
            }
        }
    }

    /// Build a microcode snapshot of the function at `address`.
    ///
    /// - Parameters:
    ///   - address: Entry address of the function to lift.
    ///   - maturity: Microcode maturity at which to snapshot.
    ///     Defaults to ``Maturity/lvars`` — the recommended choice for
    ///     MIR-style consumers.
    public static func snapshot(
        of address: Address,
        at maturity: Maturity = .lvars
    ) throws(IDAError) -> FunctionSnapshot {
        var raw: IdaxMicrocodeSnapshotHandle? = nil
        try checkStatus(
            idax_microcode_snapshot_create(address, maturity.rawValue, &raw),
            "microcode.snapshot"
        )
        guard let raw else {
            throw IDAError(category: .internal,
                           code: 0,
                           message: "microcode.snapshot returned null handle")
        }
        return FunctionSnapshot(raw)
    }
}

// MARK: - Internal conversions

private extension Microcode.Operand {
    init(raw: IdaxMicrocodeSnapshotOperand) {
        self.kind                = Microcode.OperandKind(rawValue: Int32(raw.kind)) ?? .none
        self.byteWidth           = Int(raw.byte_width)
        self.registerID          = Int(raw.register_id)
        self.secondRegisterID    = Int(raw.second_register_id)
        self.numericValue        = raw.numeric_value
        self.floatValue          = raw.float_value
        self.stackOffset         = raw.stack_offset
        self.globalAddress       = raw.global_address
        self.localVariableIndex  = Int(raw.local_variable_index)
        self.localVariableOffset = raw.local_variable_offset
        self.helperName          = borrowCString(raw.helper_name)
        self.stringLiteral       = borrowCString(raw.string_literal)
        self.blockIndex          = Int(raw.block_index)
        self.nestedInstructionID = Int(raw.nested_instruction_id)
        self.ssaVersion          = raw.has_ssa_version != 0 ? Int(raw.ssa_version) : nil
        self.operandProperties   = raw.operand_properties
    }
}

private extension Microcode.Instruction {
    init(raw: IdaxMicrocodeSnapshotInstruction) {
        self.id            = Int(raw.id)
        self.sourceAddress = raw.source_address
        self.opcode        = Int(raw.opcode)
        self.opcodeName    = borrowCString(raw.opcode_name)
        self.left          = Microcode.Operand(raw: raw.left)
        self.right         = Microcode.Operand(raw: raw.right)
        self.destination   = Microcode.Operand(raw: raw.destination)
        self.flags         = raw.flags
    }
}

private extension Microcode.Block {
    init(raw: IdaxMicrocodeSnapshotBlock) {
        self.index         = Int(raw.index)
        self.startAddress  = raw.start_address
        self.endAddress    = raw.end_address
        self.kind          = Microcode.BlockKind(rawValue: Int32(raw.kind)) ?? .none
        self.flags         = raw.flags

        if let pointer = raw.predecessor_indices, raw.predecessor_count > 0 {
            self.predecessorIndices = UnsafeBufferPointer(start: pointer,
                                                          count: raw.predecessor_count).map { Int($0) }
        } else {
            self.predecessorIndices = []
        }

        if let pointer = raw.successor_indices, raw.successor_count > 0 {
            self.successorIndices = UnsafeBufferPointer(start: pointer,
                                                       count: raw.successor_count).map { Int($0) }
        } else {
            self.successorIndices = []
        }

        if let pointer = raw.instructions, raw.instruction_count > 0 {
            self.instructions = UnsafeBufferPointer(start: pointer,
                                                    count: raw.instruction_count)
                .map { Microcode.Instruction(raw: $0) }
        } else {
            self.instructions = []
        }
    }
}

File: src/microcode.cpp
/// \file microcode.cpp
/// \brief Implementation of `ida::microcode` — post-hoc snapshot of
///        Hex-Rays microcode (mba_t/mblock_t/minsn_t/mop_t).
///
/// The snapshot is constructed by calling `gen_microcode()` to materialise a
/// fresh `mba_t` at the requested maturity, walking it once to deep-copy
/// every operand into plain owned value types, and then releasing the SDK
/// allocation. Consumers receive a self-contained `FunctionSnapshot` whose
/// lifetime is independent of the decompiler.

#include "detail/sdk_bridge.hpp"
#include <ida/microcode.hpp>
#include <ida/decompiler.hpp>

#include <hexrays.hpp>

#include <array>
#include <cstring>
#include <unordered_map>
#include <utility>

namespace ida::microcode {

namespace {

/// Translate `ida::microcode::Maturity` into the SDK enum.
mba_maturity_t to_sdk_maturity(Maturity maturity) noexcept {
    switch (maturity) {
        case Maturity::Generated:       return MMAT_GENERATED;
        case Maturity::Preoptimized:    return MMAT_PREOPTIMIZED;
        case Maturity::Locopt:          return MMAT_LOCOPT;
        case Maturity::CalledArguments: return MMAT_CALLS;
        case Maturity::Glbopt1:         return MMAT_GLBOPT1;
        case Maturity::Glbopt2:         return MMAT_GLBOPT2;
        case Maturity::Glbopt3:         return MMAT_GLBOPT3;
        case Maturity::Lvars:           return MMAT_LVARS;
    }
    return MMAT_LVARS;
}

/// Reverse mapping for round-tripping into the snapshot.
Maturity from_sdk_maturity(mba_maturity_t maturity) noexcept {
    switch (maturity) {
        case MMAT_GENERATED:    return Maturity::Generated;
        case MMAT_PREOPTIMIZED: return Maturity::Preoptimized;
        case MMAT_LOCOPT:       return Maturity::Locopt;
        case MMAT_CALLS:        return Maturity::CalledArguments;
        case MMAT_GLBOPT1:      return Maturity::Glbopt1;
        case MMAT_GLBOPT2:      return Maturity::Glbopt2;
        case MMAT_GLBOPT3:      return Maturity::Glbopt3;
        case MMAT_LVARS:        return Maturity::Lvars;
        case MMAT_ZERO:
        default:                return Maturity::Generated;
    }
}

BlockKind to_block_kind(mblock_type_t kind) noexcept {
    switch (kind) {
        case BLT_NONE: return BlockKind::None;
        case BLT_STOP: return BlockKind::Stop;
        case BLT_0WAY: return BlockKind::NoWay;
        case BLT_1WAY: return BlockKind::OneWay;
        case BLT_2WAY: return BlockKind::TwoWay;
        case BLT_NWAY: return BlockKind::NWay;
        case BLT_XTRN: return BlockKind::External;
    }
    return BlockKind::None;
}

/// Resolve a symbolic name for an `mcode_t` opcode.
///
/// SDK does not export a name table for `mcode_t`, so we maintain our own.
/// Unknown opcodes fall back to `"m_unknown_<hex>"` so the snapshot still
/// carries something useful to downstream consumers.
std::string opcode_to_string(mcode_t opcode) {
    static constexpr std::array<const char*, m_max> names = {
        "m_nop",   "m_stx",   "m_ldx",   "m_ldc",   "m_mov",   "m_neg",
        "m_lnot",  "m_bnot",  "m_xds",   "m_xdu",   "m_low",   "m_high",
        "m_add",   "m_sub",   "m_mul",   "m_udiv",  "m_sdiv",  "m_umod",
        "m_smod",  "m_or",    "m_and",   "m_xor",   "m_shl",   "m_shr",
        "m_sar",   "m_cfadd", "m_ofadd", "m_cfshl", "m_cfshr", "m_sets",
        "m_seto",  "m_setp",  "m_setnz", "m_setz",  "m_setae", "m_setb",
        "m_seta",  "m_setbe", "m_setg",  "m_setge", "m_setl",  "m_setle",
        "m_jcnd",  "m_jnz",   "m_jz",    "m_jae",   "m_jb",    "m_ja",
        "m_jbe",   "m_jg",    "m_jge",   "m_jl",    "m_jle",   "m_jtbl",
        "m_ijmp",  "m_goto",  "m_call",  "m_icall", "m_ret",   "m_push",
        "m_pop",   "m_und",   "m_ext",   "m_f2i",   "m_f2u",   "m_i2f",
        "m_u2f",   "m_f2f",   "m_fneg",  "m_fadd",  "m_fsub",  "m_fmul",
        "m_fdiv",
    };
    const int code = static_cast<int>(opcode);
    if (code >= 0 && code < static_cast<int>(names.size()))
        return names[code];
    char buf[24];
    qsnprintf(buf, sizeof(buf), "m_unknown_%02x", code & 0xff);
    return buf;
}

double decode_float_constant(const fnumber_t* fnum) noexcept {
    if (fnum == nullptr)
        return 0.0;
    // SDK keeps floats in IEEE-internal `fpvalue_t`; we surface the binary
    // representation for sizes 4 and 8 so the snapshot is usable even though
    // we cannot perform 80-bit conversion without SDK helpers.
    if (fnum->nbytes == 4) {
        float value{};
        std::memcpy(&value, fnum->fnum.w, sizeof(value));
        return static_cast<double>(value);
    }
    if (fnum->nbytes == 8) {
        double value{};
        std::memcpy(&value, fnum->fnum.w, sizeof(value));
        return value;
    }
    return 0.0;
}

struct OperandPopulationContext {
    int next_nested_id = 0;
    std::unordered_map<int, Instruction> nested;
};

Operand snapshot_operand(const mop_t& op, OperandPopulationContext& ctx);
Instruction snapshot_instruction(const minsn_t& insn, OperandPopulationContext& ctx, int id);

int register_nested(const minsn_t* nested,
                    OperandPopulationContext& ctx) {
    if (nested == nullptr)
        return -1;
    const int id = ctx.next_nested_id++;
    // Insert a placeholder first so a self-referential descent (should it
    // ever occur) cannot recurse infinitely.
    auto [it, inserted] = ctx.nested.try_emplace(id, Instruction{});
    it->second = snapshot_instruction(*nested, ctx, id);
    return id;
}

Operand snapshot_operand(const mop_t& op, OperandPopulationContext& ctx) {
    Operand snap;
    snap.byte_width        = op.size;
    snap.operand_properties = op.oprops;
    if (op.valnum != 0)
        snap.ssa_version = op.valnum;

    switch (op.t) {
        case mop_z:
            snap.kind = Operand::Kind::None;
            break;
        case mop_r:
            snap.kind        = Operand::Kind::Register;
            snap.register_id = op.r;
            break;
        case mop_n:
            snap.kind = Operand::Kind::NumericConstant;
            if (op.nnn != nullptr)
                snap.numeric_value = static_cast<std::int64_t>(op.nnn->value);
            break;
        case mop_str:
            snap.kind = Operand::Kind::StringLiteral;
            if (op.cstr != nullptr)
                snap.string_literal = op.cstr;
            break;
        case mop_d:
            snap.kind                  = Operand::Kind::NestedInstruction;
            snap.nested_instruction_id = register_nested(op.d, ctx);
            break;
        case mop_S:
            snap.kind = Operand::Kind::StackVariable;
            if (op.s != nullptr)
                snap.stack_offset = op.s->off;
            break;
        case mop_v:
            snap.kind           = Operand::Kind::GlobalAddress;
            snap.global_address = op.g;
            break;
        case mop_b:
            snap.kind        = Operand::Kind::BlockReference;
            snap.block_index = op.b;
            break;
        case mop_f:
            // Detailed call info is not surfaced in the snapshot today; only
            // the kind is recorded so downstream consumers can detect calls.
            snap.kind = Operand::Kind::CallInfo;
            break;
        case mop_l:
            snap.kind = Operand::Kind::LocalVariable;
            if (op.l != nullptr) {
                snap.local_variable_index  = op.l->idx;
                snap.local_variable_offset = op.l->off;
            }
            break;
        case mop_a:
            snap.kind = Operand::Kind::AddressOf;
            if (op.a != nullptr) {
                // mop_addr_t embeds another mop_t describing the pointed-to
                // operand. We surface the most useful identifier per kind.
                switch (op.a->t) {
                    case mop_r:
                        snap.register_id = op.a->r;
                        break;
                    case mop_v:
                        snap.global_address = op.a->g;
                        break;
                    case mop_S:
                        if (op.a->s != nullptr)
                            snap.stack_offset = op.a->s->off;
                        break;
                    case mop_l:
                        if (op.a->l != nullptr) {
                            snap.local_variable_index  = op.a->l->idx;
                            snap.local_variable_offset = op.a->l->off;
                        }
                        break;
                    default:
                        break;
                }
            }
            break;
        case mop_h:
            snap.kind = Operand::Kind::Helper;
            if (op.helper != nullptr)
                snap.helper_name = op.helper;
            break;
        case mop_c:
            snap.kind = Operand::Kind::Cases;
            break;
        case mop_fn:
            snap.kind        = Operand::Kind::FloatConstant;
            snap.float_value = decode_float_constant(op.fpc);
            break;
        case mop_p:
            snap.kind = Operand::Kind::RegisterPair;
            if (op.pair != nullptr) {
                if (op.pair->lop.t == mop_r)
                    snap.register_id = op.pair->lop.r;
                if (op.pair->hop.t == mop_r)
                    snap.second_register_id = op.pair->hop.r;
            }
            break;
        case mop_sc:
            snap.kind = Operand::Kind::Scattered;
            break;
        default:
            snap.kind = Operand::Kind::None;
            break;
    }
    return snap;
}

Instruction snapshot_instruction(const minsn_t& insn,
                                 OperandPopulationContext& ctx,
                                 int id) {
    Instruction snap;
    snap.id             = id;
    snap.source_address = static_cast<Address>(insn.ea);
    snap.opcode         = static_cast<int>(insn.opcode);
    snap.opcode_name    = opcode_to_string(insn.opcode);
    snap.flags          = static_cast<std::uint32_t>(insn.iprops);
    snap.left           = snapshot_operand(insn.l, ctx);
    snap.right          = snapshot_operand(insn.r, ctx);
    snap.destination    = snapshot_operand(insn.d, ctx);
    return snap;
}

} // namespace

// ── FunctionSnapshot::Impl ──────────────────────────────────────────────

struct FunctionSnapshot::Impl {
    Address                                            function_address{BadAddress};
    Maturity                                           maturity{Maturity::Lvars};
    std::int64_t                                       local_variables_size{0};
    std::int64_t                                       saved_registers_size{0};
    std::int64_t                                       stack_size{0};
    std::vector<Block>                                 blocks;
    std::unordered_map<int, Instruction>               nested_instructions;
    std::vector<ida::decompiler::LocalVariable>        local_variables;
};

FunctionSnapshot::FunctionSnapshot() = default;
FunctionSnapshot::~FunctionSnapshot() = default;
FunctionSnapshot::FunctionSnapshot(const FunctionSnapshot&) = default;
FunctionSnapshot& FunctionSnapshot::operator=(const FunctionSnapshot&) = default;
FunctionSnapshot::FunctionSnapshot(FunctionSnapshot&&) noexcept = default;
FunctionSnapshot& FunctionSnapshot::operator=(FunctionSnapshot&&) noexcept = default;

FunctionSnapshot::FunctionSnapshot(std::shared_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}

Address FunctionSnapshot::function_address() const noexcept {
    return impl_ ? impl_->function_address : BadAddress;
}

Maturity FunctionSnapshot::maturity() const noexcept {
    return impl_ ? impl_->maturity : Maturity::Lvars;
}

std::int64_t FunctionSnapshot::local_variables_size() const noexcept {
    return impl_ ? impl_->local_variables_size : 0;
}

std::int64_t FunctionSnapshot::saved_registers_size() const noexcept {
    return impl_ ? impl_->saved_registers_size : 0;
}

std::int64_t FunctionSnapshot::stack_size() const noexcept {
    return impl_ ? impl_->stack_size : 0;
}

const std::vector<Block>& FunctionSnapshot::blocks() const noexcept {
    static const std::vector<Block> empty;
    return impl_ ? impl_->blocks : empty;
}

bool FunctionSnapshot::empty() const noexcept {
    return !impl_ || impl_->blocks.empty();
}

Result<Instruction> FunctionSnapshot::nested_instruction(int id) const {
    if (!impl_)
        return std::unexpected(Error::not_found("Empty microcode snapshot"));
    auto it = impl_->nested_instructions.find(id);
    if (it == impl_->nested_instructions.end())
        return std::unexpected(Error::not_found(
            "No nested microinstruction with id " + std::to_string(id)));
    return it->second;
}

Result<std::vector<ida::decompiler::LocalVariable>>
FunctionSnapshot::local_variables() const {
    if (!impl_)
        return std::unexpected(Error::not_found("Empty microcode snapshot"));
    return impl_->local_variables;
}

// ── snapshot() entry point ──────────────────────────────────────────────

Result<FunctionSnapshot> snapshot(Address function_address, Maturity maturity) {
    if (function_address == BadAddress)
        return std::unexpected(Error::validation("function_address is BadAddress"));

    auto availability = ida::decompiler::available();
    if (!availability)
        return std::unexpected(availability.error());
    if (!*availability)
        return std::unexpected(Error::unsupported(
            "Decompiler not available (Hex-Rays plugin not loaded)"));

    func_t* function = get_func(static_cast<ea_t>(function_address));
    if (function == nullptr)
        return std::unexpected(Error::not_found(
            "No function at address", std::to_string(function_address)));

    mba_ranges_t ranges(function);
    hexrays_failure_t failure;
    const mba_maturity_t sdk_maturity = to_sdk_maturity(maturity);

    mba_t* mba = gen_microcode(ranges, &failure, nullptr, 0, sdk_maturity);
    if (mba == nullptr) {
        qstring description = failure.desc();
        return std::unexpected(Error::sdk(
            "gen_microcode failed",
            ida::detail::to_string(description)));
    }

    // The SDK ratchets maturity forward; honour whichever level it actually
    // produced rather than the one we requested.
    auto impl = std::make_shared<FunctionSnapshot::Impl>();
    impl->function_address     = function_address;
    impl->maturity             = from_sdk_maturity(mba->maturity);
    impl->local_variables_size = static_cast<std::int64_t>(mba->frsize);
    impl->saved_registers_size = static_cast<std::int64_t>(mba->frregs);
    impl->stack_size           = static_cast<std::int64_t>(mba->stacksize);

    OperandPopulationContext ctx;

    impl->blocks.reserve(mba->qty);
    for (int block_index = 0; block_index < mba->qty; ++block_index) {
        const mblock_t* sdk_block = mba->get_mblock(block_index);
        if (sdk_block == nullptr)
            continue;

        Block block;
        block.index         = sdk_block->serial;
        block.start_address = static_cast<Address>(sdk_block->start);
        block.end_address   = static_cast<Address>(sdk_block->end);
        block.kind          = to_block_kind(sdk_block->type);
        block.flags         = sdk_block->flags;

        block.predecessor_indices.reserve(sdk_block->npred());
        for (int p = 0; p < sdk_block->npred(); ++p)
            block.predecessor_indices.push_back(sdk_block->pred(p));

        block.successor_indices.reserve(sdk_block->nsucc());
        for (int s = 0; s < sdk_block->nsucc(); ++s)
            block.successor_indices.push_back(sdk_block->succ(s));

        for (const minsn_t* mi = sdk_block->head; mi != nullptr; mi = mi->next) {
            const int top_level_id = ctx.next_nested_id++;
            // Reserve the slot so any nested operand created during the walk
            // gets a stable, unique id that is also visible via
            // `nested_instruction()`.
            auto [it, inserted] = ctx.nested.try_emplace(top_level_id, Instruction{});
            Instruction instruction = snapshot_instruction(*mi, ctx, top_level_id);
            it->second = instruction;
            block.instructions.push_back(std::move(instruction));
        }

        impl->blocks.push_back(std::move(block));
    }

    impl->nested_instructions = std::move(ctx.nested);

    // Local variable table — only fully populated from MMAT_LVARS onward,
    // but the SDK always exposes whatever it has so far.
    impl->local_variables.reserve(mba->vars.size());
    for (std::size_t i = 0; i < mba->vars.size(); ++i) {
        const lvar_t& v = mba->vars[i];
        ida::decompiler::LocalVariable lv;
        lv.index       = i;
        lv.name        = ida::detail::to_string(v.name);
        lv.is_argument = v.is_arg_var();
        lv.width       = v.width;
        qstring type_str;
        if (v.type().print(&type_str))
            lv.type_name = ida::detail::to_string(type_str);
        else
            lv.type_name = "(unknown)";
        lv.has_user_name = v.has_user_name();
        lv.has_nice_name = v.has_nice_name();
        lv.comment       = ida::detail::to_string(v.cmt);
        if (v.is_stk_var())
            lv.storage = ida::decompiler::VariableStorage::Stack;
        else if (v.is_reg_var())
            lv.storage = ida::decompiler::VariableStorage::Register;
        else
            lv.storage = ida::decompiler::VariableStorage::Unknown;
        impl->local_variables.push_back(std::move(lv));
    }

    // The SDK hands us ownership of `mba`; release it now that we hold a
    // self-contained snapshot.
    delete mba;

    return FunctionSnapshot(std::move(impl));
}

} // namespace ida::microcode
